libmamba/src/core/curl.cpp

```diff
@@ -628,4 +669,88 @@ namespace mamba
         return static_cast<std::size_t>(numfds);
     }
 
+    CURLUHandle::CURLUHandle()
+    {
+        m_handle = curl_url(); /* get a handle to work with */
+        if (!m_handle)
+        {
+            throw std::runtime_error("Could not initiate URL parser.");
+        }
+    }
+
+    CURLUHandle::~CURLUHandle()
+    {
+        curl_url_cleanup(m_handle);
+    }
+
+    CURLUHandle::CURLUHandle(CURLUHandle&& rhs)
+        : m_handle(std::move(rhs.m_handle))
+    {
+    }
+
+    CURLUHandle& CURLUHandle::operator=(CURLUHandle&& rhs)
+    {
+        std::swap(m_handle, rhs.m_handle);
+        return *this;
+    }
+
+    void CURLUHandle::set_curl_url(const std::string& url, bool has_scheme)
+    {
+        CURLUcode uc;
+        auto curl_flags = has_scheme ? CURLU_NON_SUPPORT_SCHEME : CURLU_DEFAULT_SCHEME;
+        std::string c_url = url::unc_url(url);
+        uc = curl_url_set(m_handle, CURLUPART_URL, c_url.c_str(), static_cast<unsigned int>(curl_flags));
+        if (uc)
+        {
+            throw std::runtime_error(
+                "Could not set URL (code: " + std::to_string(uc) + " - url = " + url + ")"
+            );
+        }
+    }
+
+    namespace
+    {
+        std::map<std::string, CURLUPart> to_curlu_part{ { "CURLUPART_URL", CURLUPART_URL },
```

Comment:
I am not sure I get the motivation for this change. Why do we need this mapping from string to integers? Didn't we have a decent URL API already?

Reply:
Since this is not meant to change (apart through development), this could also be a `constexpr` function like the following. Slightly more annoying to write, but it has no global state and can be inlined / optimized away. 
```cpp
constexpr auto to_curlu_part(std::string_view name) {
    if(name == "CURLUPART_URL"){
        return CURLUPART_URL;
    } else if (...) {
       ...
   }
}
```

Reply:
But we already have encapsulation with the `get_scheme` `get_path`, ... functions?
And if we go for strings, why keep the names?

But in general, using the functions should be much preferred IMO, and then the get_part function doesn't need to be exposed at all.

Reply:
> And if we go for strings, why keep the names?

We keep them to be explicit.

> But in general, using the functions should be much preferred IMO, and then the get_part function doesn't need to be exposed at all.

Yes indeed, so the exposure here is not the problem, this mapping is more intended to avoid including specific curl header  (`curl/urlapi.h`) in url files.

Reply:
Any thoughts on this mapping? @JohanMabille @Klaim

Reply:
if it's really necessary, why not make an enum?